database2/ffldb/blockio.go

```diff
@@ -0,0 +1,771 @@
+	wc := s.writeCursor
+	wc.RLock()
+	if fileNum == wc.curFileNum && wc.curFile.file != nil {
+		obf := wc.curFile
+		obf.RLock()
+		wc.RUnlock()
+		return obf, nil
+	}
+	wc.RUnlock()
+
+	// Try to return an open file under the overall files read lock.
+	s.obfMutex.RLock()
+	if obf, ok := s.openBlockFiles[fileNum]; ok {
+		s.lruMutex.Lock()
+		s.openBlocksLRU.MoveToFront(s.fileNumToLRUElem[fileNum])
+		s.lruMutex.Unlock()
+
+		obf.RLock()
+		s.obfMutex.RUnlock()
+		return obf, nil
+	}
+	s.obfMutex.RUnlock()
+
+	// Since the file isn't open already, need to check the open block files
+	// map again under write lock in case multiple readers got here and a
+	// separate one is already opening the file.
+	s.obfMutex.Lock()
+	if obf, ok := s.openBlockFiles[fileNum]; ok {
+		obf.RLock()
+		s.obfMutex.Unlock()
+		return obf, nil
+	}
+
+	// The file isn't open, so open it while potentially closing the least
+	// recently used one as needed.
+	obf, err := s.openFileFunc(fileNum)
+	if err != nil {
+		s.obfMutex.Unlock()
+		return nil, err
+	}
+	obf.RLock()
+	s.obfMutex.Unlock()
+	return obf, nil
+}
+
+// writeData is a helper function for writeBlock which writes the provided data
+// at the current write offset and updates the write cursor accordingly.  The
+// field name parameter is only used when there is an error to provide a nicer
+// error message.
+//
+// The write cursor will be advanced the number of bytes actually written in the
+// event of failure.
+//
+// NOTE: This function MUST be called with the write cursor current file lock
+// held and must only be called during a write transaction so it is effectively
+// locked for writes.  Also, the write cursor current file must NOT be nil.
+func (s *blockStore) writeData(data []byte, fieldName string) error {
+	wc := s.writeCursor
+	n, err := wc.curFile.file.WriteAt(data, int64(wc.curOffset))
+	wc.curOffset += uint32(n)
+	if err != nil {
+		str := fmt.Sprintf("failed to write %s to file %d at "+
+			"offset %d: %v", fieldName, wc.curFileNum,
+			wc.curOffset-uint32(n), err)
+		return makeDbErr(database.ErrDriverSpecific, str, err)
+	}
+
+	return nil
+}
+
+// writeBlock appends the specified raw block bytes to the store's write cursor
+// location and increments it accordingly.  When the block would exceed the max
+// file size for the current flat file, this function will close the current
+// file, create the next file, update the write cursor, and write the block to
+// the new file.
+//
+// The write cursor will also be advanced the number of bytes actually written
+// in the event of failure.
+//
+// Format: <network><block length><serialized block><checksum>
+func (s *blockStore) writeBlock(rawBlock []byte) (blockLocation, error) {
+	// Compute how many bytes will be written.
+	// 4 bytes each for block network + 4 bytes for block length +
+	// length of raw block + 4 bytes for checksum.
+	blockLen := uint32(len(rawBlock))
+	fullLen := blockLen + 12
```

Comment:
It's not the same thing.  `blockLocSize` refers to what's kept in the database metadata record (in particular the block file number, file offset, and full length).  This is referring to what is physically going to the disk in the block files.  As the comment calls out it's network + block length + block + checksum.